drivers/can/can_sam.c

```diff
@@ -0,0 +1,260 @@
+		.prop_ts1 = DT_INST_PROP_OR(inst, prop_seg, 0) +               \
+			    DT_INST_PROP_OR(inst, phase_seg1, 0),              \
+		.ts2 = DT_INST_PROP_OR(inst, phase_seg2, 0),                   \
+		.bus_speed_data = DT_INST_PROP(inst, bus_speed_data),          \
+		.sjw_data = DT_INST_PROP(inst, sjw_data),                      \
+		.sample_point_data =                                           \
+			DT_INST_PROP_OR(inst, sample_point_data, 0),           \
+		.prop_ts1_data = DT_INST_PROP_OR(inst, prop_seg_data, 0) +     \
+				 DT_INST_PROP_OR(inst, phase_seg1_data, 0),    \
+		.ts2_data = DT_INST_PROP_OR(inst, phase_seg2_data, 0),         \
+		.tx_delay_comp_offset =                                        \
+			DT_INST_PROP(inst, tx_delay_comp_offset)               \
+	}
+#else /* CONFIG_CAN_FD_MODE */
+#define CAN_SAM_MCAN_CFG(inst)                                                 \
+	{                                                                      \
+		.can = (struct can_mcan_reg *)DT_INST_REG_ADDR(inst),          \
+		.bus_speed = DT_INST_PROP(inst, bus_speed),                    \
+		.sjw = DT_INST_PROP(inst, sjw),                                \
+		.sample_point = DT_INST_PROP_OR(inst, sample_point, 0),        \
+		.prop_ts1 = DT_INST_PROP_OR(inst, prop_seg, 0) +               \
+			    DT_INST_PROP_OR(inst, phase_seg1, 0),              \
+		.ts2 = DT_INST_PROP_OR(inst, phase_seg2, 0),                   \
+	}
+#endif /* CONFIG_CAN_FD_MODE */
+
+#define CAN_SAM_CFG_INST(inst)                                                 \
+static const struct can_sam_config can_sam_cfg_##inst = {                      \
+	.pmc_id = DT_INST_PROP(inst, peripheral_id),                           \
+	.pin_list = {                                                          \
+		ATMEL_SAM_DT_INST_PIN(inst, 0),                                \
+		ATMEL_SAM_DT_INST_PIN(inst, 1)                                 \
+	},                                                                     \
+	.config_irq = config_can_##inst##_irq,                                 \
+	.mcan_cfg = CAN_SAM_MCAN_CFG(inst)                                     \
+};
+
+#define CAN_SAM_DATA_INST(inst)                                                \
+static struct can_sam_data can_sam_dev_data_##inst;
+
+#define CAN_SAM_DEVICE_INST(inst)                                              \
+DEVICE_DT_INST_DEFINE(inst, &can_sam_init, NULL,                               \
+		      &can_sam_dev_data_##inst, &can_sam_cfg_##inst,           \
+		      POST_KERNEL, CONFIG_CAN_INIT_PRIORITY,                   \
+		      &can_api_funcs);
+
+#define CAN_SAM_INST(inst)                                                     \
+CAN_SAM_IRQ_CFG_FUNCTION(inst)                                                 \
+CAN_SAM_CFG_INST(inst)                                                         \
+CAN_SAM_DATA_INST(inst)                                                        \
+CAN_SAM_DEVICE_INST(inst)
```

Comment:
Lots of macros going on here, some just for defining a single line. I would collapse as much as this into one macro as we tend to do elsewhere in Zephyr (see e.g. the NXP MCUX M_CAN frontend driver).

Reply:
I think it is much more readable and easier to debug this way.
Most of the macros are more than one line. To be consistent, I added macros for all definitions.